packages/commonwealth/server/util/versioning.ts

```diff
@@ -0,0 +1,38 @@
+import { AddressInstance } from '@hicommonwealth/model';
+import moment from 'moment';
+
+export function addVersionHistory(
+  oldVersionHistory: string[],
+  body: string,
+  address: AddressInstance,
+): { latestVersion: string; versionHistory: string[] } {
+  // update version history
+  let latestVersion;
+  try {
+    latestVersion = JSON.parse(oldVersionHistory[0]).body;
+  } catch (err) {
+    console.log(err);
+  }
+
+  let parsedBody;
+  try {
+    parsedBody = JSON.parse(decodeURIComponent(body)).ops[0].insert;
+  } catch (error) {
+    // If parsing fails, or the property doesn't exist, assign the original body
+    parsedBody = decodeURIComponent(body);
+  }
+
+  if (parsedBody !== latestVersion) {
+    const recentEdit = {
+      timestamp: moment(),
+      author: address,
```

Comment:
The old data structure for `author` was `MinimumProfile` while the new one is `Address`. Older content still contains version history entries with the old format and when edited, will contain both the old and new format in the version history array.

All entries show up in the UI, but only the newer entries render a profile user.

I think we should do something about data consistency by adding a migration– either delete all old version history or migrate them to the new structure?

Additionally, there should be better type safety for the thread/comment `version_history` property, given that it *is* structured data. Right now it's a string array. Maybe make it a JSON/B column so there's at least some basic JSON validation?

Reply:
> The old data structure for author was MinimumProfile while the new one is Address

The idea of a MinimumProfile only exists on the frontend. There was no old datastructure for the backend. We had:
1. create_thread_comment: {timestamp, body}
2. create_thread: {timestamp, author: Address, body}
3. update_comment: {timestamp, body}
4. update_thread: {timestamp, author: (address string only), body}

This PR basically unifies the typing because we need profile ids on the front end in order to display profile information related to the edit.

> I think we should do something about data consistency by adding a migration– either delete all old version history

I wanted to avoid this approach because I feel like we shouldn't delete old inconsistent data as it still has some useful information that we can display (just doesn't display the profile/address that made the edit).

> or migrate them to the new structure?

I wanted to avoid this approach as well because there will be too many edge cases. I am not even sure the old version_hisotry structure is consistent across time for each of the 4 handlers that modify it. So both the development time will increase a ton, as well as I am not sure we would even have enough information to unify all the version history entries (Example, what if you have a thread collaboration with no address? How do you tell which profile made the edit?)

> Additionally, there should be better type safety for the thread/comment version_history property

I agree with this. But the issue is that in order to enforce the type at the database level, we need to first write the migration to change all the old version_histories data to adhere to the type. The problem is, as mentioned in the earlier section, I think the only feasible way is to drop all the old version_history data and start fresh, but then we would be losing valuable incomplete information we could use on the front end.

Reply:
There's a 4th option where we can still keep the old data, but move it to another column so that the `version_history` column can be structured. That way the backend code can work with structured data and the UI can have access to both new and old data to render as it sees fit.

Reply:
Maybe it would be best to try option 2 first, and then if it fails we can try option 4. I think data integrity is important, but it would get confusing from a development perspective having 2 fields.

Reply:
Ok I don't know if we still want to go forward with this. Basically the field will not have the type, but instead it will be a pgsql CONSTRAINT on the array of jsons (Basically parses each json and makes a bunch of checks). This will make insertion super slow i think since it would perform this validation on every version_history in the array, on every insert/update. For example, if you have 5 version histories, you add a new one, it performs this check on all 5 previous ones as well as the new one.

Maybe I can get @timolegros opinion here since maybe there is some other way to do it in pgsql that I am not seeing.

Reply:
My concern was first and foremost using better typing in code. So going from `string[]` to `ValidationHistory[]` with the column being JSON/B.

DB-level validation on insert is nice, but build-time type checks is what improves the dev experience, and is likely sufficient as long as we use the ORM for inserts/updates.

The old data is tech debt. Right now that tech is holding back the BE from having adequate typing, and is increasing complexity on the FE in order to render the various data structures.

With option 4, that tech debt would be mostly transferred to the FE. The BE would never be concerned about old data aside from passing it to the FE.

In the future, we may decide to drop the old data or finally tackle a migration for it– but either way we can move forward with better typing.

Reply:
Maybe we can discuss it during platform chat and then write up our conclusion here? As this discussion is probably easier to resolve in person.

As a side note I tried converting the column types with something like:

```sql
-- create version_history_new column

UPDATE "Threads" SET version_history_new = (
    SELECT jsonb_agg(unnest::jsonb) 
    FROM unnest(version_history)
);

-- remove version_history and rename new_version_history to it

-- perform the same steps for comments
```

But this migration takes about 50 seconds on my computer. These migrations usually take 4-5 times as long on our heroku dyno so I don't know how feasible option 2 would be, which would leave us with option 4.

Reply:
Additionally, we should convert the type of `version_history` from `TEXT[]` to `JSONB` (without any DB level constraints on the schema).

Reply:
Actually, come to think of it. If we convert the column to `JSONB` all we need is the proper union type in the Zod schema to achieve proper type safety.

Reply:
For posterity. In platform chat we concluded that we would migrate the version_history text[] to a new table. This way we would get type safety as well as referential integrity + better scaling.

But for now, since we have a lot on the plate, I was wondering if that could be performed in a followup PR and instead we just get this out now since it works, and then in the followup PR migrate these to their own table.